### ISSUE TITLE
Signal suggestion

### ISSUE THREAD

**Issue by git2013vb** [godotengine/godot#14381]
Can we inherit custom signals  in children nodes, please?
Or alternately  make a "in memory stack" where i have all signals with their paths related with their current position in node tree.

These paths will be dynamically update if  any node will change position.

Thank you.

**Comment by vnen** [comment]
It doesn't inherit signals? I wonder why not, the built-in signals are inherited just fine.

**Comment by git2013vb** [comment]
Was a surprise to me too. I didn't check if the 2.4 had same behavior.

**Comment by MarianoGnu** [comment]
no, signals are inherited by class inheritance, not by parent/child relationship imagine you have an sprite children of an area2d, do you expect to EMIT a body_enter signal? that doesn't make any sense. also if it does, how many area2d could be in highter levels of the scene?

**Comment by git2013vb** [comment]
I talk about custom signals.

**Comment by MarianoGnu** [comment]
that would only lead to confusion, i'm not up to it, breaks encapsulation, class declaration and class inheritance.
If you want to emit you parent's signal just call get_parent().emit_signal(...)

**Comment by git2013vb** [comment]
confusion and flexibility are 2 side of same medal. In other hand we talk about a powerful feature where we can avoid to hard-code node-path in code.. At least in children.
Unique naming convention and we can fire a signal without make any modification in gdscript even if we move nodes in tree - at least if we keep that node in lower level in compared to node' signal.
like this:

NodeBase:
signal my_signal

....

...
somewhere in tree node
my_node:
connect("my_signal", "my_other_node", "my_function", params[])
.....

**Comment by MarianoGnu** [comment]
"somewhere in tree node" yeah, that kind of programing is very deterministic :/
you dont even know wich nodes or how many nodes will that call connect. Using groups is much safer and is deterministic since you most probably know what nodes are on each group

**Comment by vnen** [comment]
Oh, I misunderstood the request. I thought you were talking about class inheritance. The child nodes in a scene tree aren't supposed to "inherit" the signals, this would lead to confusion.

**Comment by git2013vb** [comment]
I don't understand. Rules are made to avoid confusion.
Is not that I cannot make a mess in any language programming if I don't follow rules.

In my opinion I would be nice/powerful to see and use any signals in my parent tree. That's it.

**Comment by MarianoGnu** [comment]
The Rule you propose is confusing, signals are determined by class, of you emit a signal "my_custom_signal" that is not exposed in reference, and is not declared on the script neighter other scripts extended by the current one, then how are you suposed to know who will emit this? maybe you dont even have a parent on the tree with this signal. What happens then? The confusion comes because:
- If you work in team, other members might not know what you intended to do
- An object wich you didn't want to emit the signal end up emmiting it because it was on the tree
- what happens if you are an Area2D node and you have given your parent sprite an "enter_body" signal?
- if you emit the signal from a child, who do you connect to? to the parent or the child?

**Comment by git2013vb** [comment]
I have no a clear view  of the current code. I'm aware that.

But I have to explain better what I intend with an use case:
We pretend to have a tree like this one (as the code is now)
Node->Node2D->Sprite
------->--------->Camers2D
------->--------->CollisionShape2D
------->Control->CanvasLayer->Label

What I meant was: I have a script in "Node" called "Manager". Is scope will be to manager of my game.
 
So my idea was to "define" signals in that script. And from another node in his children eg: "Label" I will use it to run some logic in "manager" script without need to use "get_node()" function... 
Why? 

Because If I need to use move/use that "Label" elsewhere or in another scene where I have no visibility of the "Node" it will be easy to build my "sub-sub-scene" without thinking: "And now what? The path that i think to put here is the right one?". Because I was thinking at some point where I can build a sub.sub-scene in GDscript . and will be difficult to me keep track the whole picture.

In short I think if there was something like that the decouple between different scene will be better.  I prefer to reuse a scene in other context instead to forced to reinvent the wheel again. 
I understand we have not enough documentation to help us to use in the right way Godot. But I think this feature will be useful.

Apologize if my English is not clear :)) I try my best.

**Comment by vnen** [comment]
@git2013vb ever heard of the XY problem? You are defending a possible solution instead of declaring what your problem really is. The scene tree is the hardest part to decouple, but it's not impossible, you just need to carefully use signals and pass NodePaths around. I believe you are overthinking this, you don't need to decouple *everything*. And you shouldn't design it around potential changes, it's easier to just make it work and then refactor when you need to expand.

If you really really need to, you can use [propagate_call](http://docs.godotengine.org/en/latest/classes/class_node.html#class-node-propagate-call) and [propagate_notification](http://docs.godotengine.org/en/latest/classes/class_node.html#class-node-propagate-notification) to pass info down the tree.

**Comment by git2013vb** [comment]
Thank you :) I will take a look .

**Comment by bojidar-bg** [comment]
Is this still a wanted feature?

Also, if it is, I think `get_owner().emit_signal()` could be more useful than `get_parent().emit_signal()` so that one can rearrange the scene without breaking things randomly.
Also, keep in mind that you can connect a signal to `emit_signal`, in order to "rebroadcast" a signal. If you do that via the scene tree, the scripts involved only need to declare the signal with `add_user_signal()` or the `signal` keyword (and you get to rearrange things at will).

**Comment by git2013vb** [comment]
Yes thanks

**Comment by git2013vb** [comment]
Hi,
After been used to use signals as are now I think is not anymore do what I suggested here. Also with the news about 4.0 I think is ok as is now ( if I need something different - that is no so different -  I can still use events in C#) 
Tank you all. Ill close this :)
